### PR TITLE
Add support for force online metadata update

### PR DESCRIPTION
**Pull request by MVrachev**
Add support for force online metadata update

Signed-off-by: Martin Vrachev <martin.vrachev@broadcom.com>

### CODE REVIEW

**Comment by MVrachev** [comment]
It's best to test this together with https://github.com/repository-service-tuf/repository-service-tuf-api/pull/513

**Comment by MVrachev** [comment]
@kairoaraujo it's ready for a review.

**Comment by MVrachev** [comment]
I invited @fsavoia and @ivanayov for a review.

File: repository_service_tuf_worker/repository.py [hunk 1]
```diff
@@ -1108,6 +1112,110 @@ def bump_online_roles(self, force: Optional[bool] = False) -> bool:
...
+            )
+        if len(payload["roles"]) < 1:
+            return self._task_result(
+                TaskName.FORCE_ONLINE_METADATA_UPDATE,
+                True,
+                {
+                    "message": "No online metadata roles given",
+                },
+            )
+
```
Comment (kairoaraujo) [inline repository_service_tuf_worker/repository.py:1173]: In the API, [we allow users to submit offline roles without checking it (example: Root)](https://github.com/repository-service-tuf/repository-service-tuf-api/blob/5594c1684064705c12f660ed68858c82636ba154/repository_service_tuf_api/metadata.py)


In the Worker, we also don't verify it. It will return a task SUCCESS with status `True`.
I suggest we verify if the role

Note:  We can verify Root and Targets now. We have in the settings/redis [a setting if the `Targets` is offline or online key](https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/0463b088095b0f8a98be91b9f80591f7d8ffa088/repository_service_tuf_worker/repository.py#L493-L495) -- for future feature that allows offline targets.

**Review by kairoaraujo — CHANGES_REQUESTED** [review]
The feature is nice. Thank you @MVrachev.
I'm adding one specific comment.

File: repository_service_tuf_worker/repository.py (hunk 1 above)
Comment (MVrachev) [reply repository_service_tuf_worker/repository.py:1173]: I changed it and now we only allow online roles and specifically check if targets is an online role.

**Comment by MVrachev** [comment]
Ready for another review.

**Comment by kairoaraujo** [comment]
I'm reviewing it now

File: repository_service_tuf_worker/repository.py [hunk 2]
```diff
@@ -61,7 +61,13 @@ class Roles(enum.Enum):
 
     @staticmethod
     def online_roles() -> List[str]:
-        return [Targets.type, Snapshot.type, Timestamp.type, "bins"]
+        online_roles = [Targets.type, Snapshot.type, Timestamp.type, "bins"]
+        settings_repository = get_worker_settings()
+        settings_repository.reload()
+        if not settings_repository.get_fresh("TARGETS_ONLINE_KEY", True):
+            online_roles = [Snapshot.type, Timestamp.type, "bins"]
```
Comment (ivanayov) [inline repository_service_tuf_worker/repository.py:68]: Shouldn't bins be excluded as well?
I might be missing some details from the workflow - if so, please ignore to not waste time with the comment. 
Otherwise, given that bins are delegates and `TARGETS_ONLINE_KEY` is `False`, it looks incorrect to have them in the online roles.

File: repository_service_tuf_worker/repository.py [hunk 3]
```diff
@@ -1078,8 +1085,11 @@ def bump_online_roles(self, force: Optional[bool] = False) -> bool:
                 expire (`self._hours_before_expire`)
         """
         logging.debug(f"Configured timeout: {self._timeout}")
-        if self._settings.get_fresh("BOOTSTRAP") is None:
-            logging.info("[automatic_version_bump] No bootstrap, skipping...")
+        bootstrap = self._settings.get_fresh("BOOTSTRAP")
+        if bootstrap is None or "pre-" in bootstrap or "signing-" in bootstrap:
+            logging.info(
+                "[automatic_version_bump] Bootstrap not completed, skipping..."
```
Comment (ivanayov) [inline repository_service_tuf_worker/repository.py:1091]: "No bootstrap or not completed, skipping"?

File: repository_service_tuf_worker/repository.py [hunk 4]
```diff
@@ -1108,6 +1118,104 @@ def bump_online_roles(self, force: Optional[bool] = False) -> bool:
...
+                )
+
+            status_lock_targets = True
+        except redis.exceptions.LockNotOwnedError:
+            # The LockNotOwnedError happens when the task exceeds the timeout,
+            # and another task owns the lock.
+            # If the task time out, the lock is released. If it doesn't finish
+            # properly, it will raise (fail) the task. Otherwise, the ignores
+            # the error because another task didn't lock it.
+            if status_lock_targets is False:
```
Comment (ivanayov) [inline repository_service_tuf_worker/repository.py:1197]: It will always be `False` as the exception is thrown prior ` status_lock_targets = True`.

File: repository_service_tuf_worker/repository.py (hunk 2 above)
Comment (MVrachev) [reply repository_service_tuf_worker/repository.py:68]: Bins could use the online key. 
Even though targets could be offline in the future this doesn't mean we are required to use offline keys for all its delegations. The security properties coming from targets utilizing an offline key wouldn't be lost as when adding, removing or downloading artifact information all requires going through `targets` first which will be cryptographically signed by an offline key.

File: repository_service_tuf_worker/repository.py (hunk 3 above)
Comment (MVrachev) [reply repository_service_tuf_worker/repository.py:1091]: I think `Bootstrap not completed` could be interpreted by both:
- not started at all
- not signed with threshold of keys

It's good that the user have `GET /api/v1/bootstrap` to understand that bootstrap is in the process of signing: https://github.com/repository-service-tuf/repository-service-tuf-api/blob/277d73b1610a5ec96c4af4d3ccbc8bdbfc69bcf9/repository_service_tuf_api/bootstrap.py#L128
and `GET /api/v1/metadata` to get the exact content of this bootstrap.

In conclusion, I don't we need to change the message to a longer now.

**Review by kairoaraujo — COMMENTED** [review]
Is it better to merge custom delegated roles and rebase them?

**Comment by MVrachev** [comment]
@kairoaraujo ready to test together with https://github.com/repository-service-tuf/repository-service-tuf-api/pull/513
Support for pushing a new version of custom target delegation is there.

File: repository_service_tuf_worker/repository.py [hunk 5]
```diff
@@ -1208,6 +1221,116 @@ def bump_online_roles(self, force: Optional[bool] = False) -> bool:
 
         return True
 
+    def _run_force_online_metadata_update(self, roles: List[str]) -> List[str]:
```
Comment (kairoaraujo) [inline repository_service_tuf_worker/repository.py:1224]: I'm wondering if the `metadata_update` is the best name here, as it confuses our process of updating the metadata.

**Review by kairoaraujo — APPROVED** [review]

File: repository_service_tuf_worker/repository.py (hunk 5 above)
Comment (MVrachev) [reply repository_service_tuf_worker/repository.py:1224]: Well it is actually metadata update...
Hard to get a name correctly...
